File: solutions/go/need-for-speed/1/need_for_speed.go
package speed

// define the 'Car' type struct
type Car struct {
    battery      int
    batteryDrain int
    speed        int
    distance     int
}

// NewCar creates a new remote controlled car with full battery and
// given specifications.
func NewCar(speed, batteryDrain int) Car {
    return Car{
        battery:      100,
        batteryDrain: batteryDrain,
        speed:        speed,
        distance:     0,
    }
}

// define the 'Track' type struct
type Track struct {
    distance int
}

// NewTrack creates a new track
func NewTrack(distance int) Track {
    return Track {
        distance: distance,
    }
}

// Drive drives the car one time. If there is not enough battery to
// drive one more time,
// the car will not move.
func Drive(car Car) Car {
    if car.battery >= car.batteryDrain {
        car.distance += car.speed
        car.battery  -= car.batteryDrain
    }
    return car
}

// CanFinish checks if a car is able to finish a certain track.
func CanFinish(car Car, track Track) bool {
    return car.battery / car.batteryDrain * car.speed >= track.distance
}

File: solutions/java/need-for-speed/1/src/main/java/NeedForSpeed.java
class NeedForSpeed {
    private int battery  = 100;
    private int distance =   0;
    private int speed;
    private int batteryDrain;

    NeedForSpeed(int speed, int batteryDrain) {
        this.speed        = speed;
        this.batteryDrain = batteryDrain;
    }

    public boolean batteryDrained() {
        return this.battery < this.batteryDrain;
    }

    public int distanceDriven() {
        return this.distance;
    }

    public void drive() {
        if ( this.battery >= this.batteryDrain ) {
            this.distance += this.speed;
            this.battery  -= this.batteryDrain;
        }
    }

    public static NeedForSpeed nitro() {
        return new NeedForSpeed(50, 4);
    }
}

class RaceTrack {
    private int distance;

    RaceTrack(int distance) {
        this.distance = distance;
    }

    public boolean canFinishRace(NeedForSpeed car) {
        while ( ! car.batteryDrained() ) {
            car.drive();
        }
        return car.distanceDriven() >= this.distance;
    }
}
